src/launcher/plugin.ts

```diff
@@ -0,0 +1,266 @@
+      launcherModel.add(args['name'] as string, args['clickCallback'] as any);
+    }
+  });
+
+
+  app.commands.addCommand('jupyterlab-launcher:show', {
+    label: 'Show Launcher',
+    execute: () => {
+      if (!launcherWidget.isAttached) {
+        app.shell.addToLeftArea(launcherWidget);
+      } else {
+        app.shell.activateMain(launcherWidget.id);
+      }
+    }
+  });
+
+  palette.addItem({
+    command: 'jupyterlab-launcher:show',
+    category: 'Help'
+  });
+
+  app.shell.addToLeftArea(launcherWidget);
+}
+
+
+/**
+ * Simple encapsulation of name and callback of launcher entries.
+ */
+class LauncherItem {
+
+  readonly name: string;
+ 
+  readonly clickCallback: () => void;
+ 
+  constructor(name: string, clickCallback: () => void) {
+    this.name = name;
+    this.clickCallback = clickCallback;
+  }
+}
+
+
+/**
+ * LauncherModel keeps track of the path to working directory and has a list of
+ * LauncherItems.
+ */
+class LauncherModel extends VDomModel {
+  items: LauncherItem[] = [];
+  path: string = 'home';
+  /**
+   * Convenience method to add a launcher with a given name and callback.
+   *
+   * This keeps plugins who wish to register themselves in the Launcher from
+   * having to import both LauncherItems and LauncherItem.
+   */
+  add(name: string, clickCallback: () => void) : IDisposable {
+    return this.addItem(new LauncherItem(name, clickCallback));
+  }
+
+  /**
+   * Add a new launcher and trigger re-render event for parent widget.
+   *
+   * returns a Disposable which can be called to remove this new item from
+   * the Launcher and trigger another re-render event.
+   */
+  addItem(item: LauncherItem) : IDisposable {
+    this.items.push(item);
+    this.stateChanged.emit(void 0);
+
+    return new DisposableDelegate(() => {
+      // Remove the item form the list of items.
+      var index = this.items.indexOf(item, 0);
+      if (index > -1) {
+          this.items.splice(index, 1);
+          this.stateChanged.emit(void 0);
+      }
+    });
+  }
+  
+  setDir(path: string) : void {
+    this.path = path;
+    this.stateChanged.emit(void 0);
+  }
+}
+
+
+/**
+ * A virtual-DOM-based widget for the Launcher.
+ */
+class LauncherWidget extends VDomWidget<LauncherModel> {
+
+  constructor() {
+    super();
+    this.addClass(LAUNCHER_CLASS);
+  }
+
+  protected render(): VNode | VNode[] {
+    let children : VNode[] = [];
+
+    for (let item of this.model.items) {
+      let imgName = item.name.replace(' ', '');
+      let img = h.span({className: `jp-Image${imgName} ` + IMAGE_CLASS});
+      let text = h.span({className:  TEXT_CLASS }, item.name);
+
+      let column = h.div({
+        className: COLUMN_CLASS,
+        'onclick': item.clickCallback
+      }, [img, text])
+      children.push(column);
+    }
+
+    let folderImage = h.span({ className: FOLDER_CLASS });
+    let p = this.model.path;
+    let pathName = p.length ? 'home > ' + p.replace('/', ' > ') : 'home';
```

Comment:
It actually behaves the way I would have expected for me. I added a `console.log` to see the value of `p`, and I took a screenshot of navigating a few layers deep:

<img width="340" src="https://cloud.githubusercontent.com/assets/159529/18977428/1e56d774-86b2-11e6-8463-e6c36782616f.png">